lib/internal/Magento/Framework/Cache/LockGuardedCacheLoader.php

```diff
@@ -93,6 +93,44 @@ public function lockedLoadData(
         return $cachedData;
     }
 
+    /**
+     * Loads cached data in non blocking way
+     *
+     * When data is not cached it will try to grab a lock.
+     * If lock is not obtainable it just returns back uncached data,
+     * so connection is not going into cyclic deadlock
+     *
+     * @param string $lockName
+     * @param callable $dataLoader
+     * @param callable $dataCollector
+     * @param callable $dataSaver
+     * @return mixed
+     */
+    public function nonBlockingLockedLoadData(
+        string $lockName,
+        callable $dataLoader,
+        callable $dataCollector,
+        callable $dataSaver
+    ) {
+        $cachedData = $dataLoader();
+
+        if ($cachedData !== false) {
+            return $cachedData;
+        }
+
+        $cachedData = $dataCollector();
+
+        if ($this->locker->lock($lockName, $this->lockTimeout / 1000)) {
```

Comment:
Hi @IvanChepurnyi 
One of the goals of the previous implementation is to reduce the load on the system due to the fact that only one client generates data. 
One write to cache is atomic for most cache adapters (at least for redis) and does not  make sense to add additional locks on top of it.

Reply:
@kandy how this introduces more than one lock?

Reply:
I mean lock at all in Magento (in additional to implement in Redis server)

Reply:
@kandy I fixed the issue with storing locked state before reading uncached data. It resolves an issue for continuous write when under high load. Redis atomic write will not save from it, as a problem is not in cache backend, but in time difference between the generation of the data on Magento side and actual write operation. When one of the processes generates data other processes still access uncached versions and schedule writes too. Good catch on the lock method forgot to add a test case for it to implement a complete solution.